designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,274 @@
+
+So we will change the default rule schema to `schema: []` to enforce that no rule options are passed to rules that do not specify a schema.
+
+This has the desirable consequence of requiring rules with options to begin specifying their schema if they did not already.
+
+Making ESLint's default behavior more strict like this goes along with many recent changes to tighten validation (i.e. more [RuleTester class validation in ESLint 7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#rule-tester-strict), [increased rule configuration validation in ESLint 6](https://eslint.org/docs/user-guide/migrating-to-6.0.0#rule-config-validating), etc) to improve rule quality and usability and reduce the chance of mistakes.
+
+And in addition to the obvious benefits of increased rule schema usage such as reduced chance of silent mistakes/bugs, being able to rely on having rule schemas available could unlock new features/improvements:
+
+- Improved/auto-generated documentation regarding rule options
+- TypeScript-style auto-complete/IntelliSense when configuring a rule
+
+There is an existing, third-party lint rule [eslint-plugin/require-meta-schema](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) to enforce that rules have schemas.
+
+### Motivation for requiring object-style rules
+
+Supporting two rule formats requires increased complexity and maintenance burden inside ESLint, as well as in third-party tools like [eslint-plugin-eslint-plugin](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin) that need to be able to handle both rule formats.
+
+Object-style rules were introduced and function-style rules were deprecated in early 2016 (?). Over five years has elapsed since then, and we don't want to support both formats forever.
+
+Object-style rules are clearly superior in terms of configuration thanks to the `meta` object they include. Function-style rules are also blocked from accessing valuable features like [autofixing](https://eslint.org/docs/8.0.0/user-guide/migrating-to-8.0.0#rules-require-metafixable-to-provide-fixes) and suggestions.
+
+There is an existing, third-party lint rule [eslint-plugin/prefer-object-rule](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-object-rule.md) to enforce and autofix that rules are written with the object-style.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The two proposals in this RFC should be implemented separately.
+
+### Detailed design for requiring schemas
+
```

Comment:
Note that these changes will need to be made in the flat config files as we aren’t making any further changes to the eslintrc format.